### PR TITLE
WIP elasticsearch boosting

### PR DESCRIPTION
**Pull request by bderusha**
[ECOM-5314](https://openedx.atlassian.net/browse/ECOM-5314) Boosting

FYI @clintonb @rlucioni

### CODE REVIEW

File: course_discovery/apps/edx_haystack_extensions/backends.py [hunk 1]
```diff
@@ -28,16 +28,119 @@ def build_search_kwargs(self, *args, **kwargs):
...
+            # Date boost upcoming courses
+            {
+                'filter': {
+                    'exists': {
+                        'field': 'start'
+                    }
+                },
+                'linear': {
+                    'start': {
+                        'origin': '2016-08-19',
```
Comment (clintonb) [inline course_discovery/apps/edx_haystack_extensions/backends.py]: 'now' is a valid value here.

**Comment by clintonb** [comment]
I played around with this branch locally for a few minutes. This is neat!!!

**Comment by bderusha** [comment]
@clintonb @rlucioni Please take a look

File: course_discovery/apps/api/models.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from django.utils.translation import ugettext_lazy as _
+from jsonfield.fields import JSONField
+from solo.models import SingletonModel
+
+
+class ElasticsearchBoostConfig(SingletonModel):
+    """
+    Model used to store default program configuration.
```
Comment (clintonb) [inline course_discovery/apps/api/models.py]: This docstring is invalid.

File: course_discovery/apps/api/models.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from django.utils.translation import ugettext_lazy as _
+from jsonfield.fields import JSONField
+from solo.models import SingletonModel
+
+
+class ElasticsearchBoostConfig(SingletonModel):
```
Comment (clintonb) [inline course_discovery/apps/api/models.py]: This model fits better in `edx_haystack_extensions`.

File: course_discovery/apps/edx_haystack_extensions/backends.py [hunk 4]
```diff
@@ -34,10 +36,21 @@ def build_search_kwargs(self, *args, **kwargs):
             'auto_generate_phrase_queries': True,
         }
 
-        if search_kwargs['query'].get('filtered', {}).get('query', {}).get('query_string'):
-            search_kwargs['query']['filtered']['query']['query_string'] = simple_query
+        # https://www.elastic.co/guide/en/elasticsearch/reference/1.7/query-dsl-function-score-query.html
+        function_score_config = ElasticsearchBoostConfig.get_solo().function_score
```
Comment (clintonb) [inline course_discovery/apps/edx_haystack_extensions/backends.py:40]: What happens if this doesn't exist?

File: course_discovery/apps/api/models.py (hunk 3 above)
Comment (clintonb) [inline course_discovery/apps/api/models.py]: Additionally, you can create the first instance by adding a migration to `edx_catalog_extensions`.

**Comment by clintonb** [comment]
Tests?

File: course_discovery/apps/edx_haystack_extensions/backends.py (hunk 4 above)
Comment (bderusha) [reply course_discovery/apps/edx_haystack_extensions/backends.py:40]: `get_solo` creates a default if one doesn't exist yet

File: course_discovery/apps/edx_haystack_extensions/models.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    # }
+
+    function_score = JSONField(
+        verbose_name=_('Function Score'),
+        help_text=_('JSON string containing an elasticsearch function score config.'),
+        null=False,
+        blank=False,
+        default={
+            "functions": [],
+            "boost": 5.0,
```
Comment (clintonb) [inline course_discovery/apps/edx_haystack_extensions/models.py]: Please add a comment explaining why we are boosting by default. Will this even have an effect if _everything_ is boosted?

File: course_discovery/apps/edx_haystack_extensions/models.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    #     "weight": 100.0,
+    # }
+
+    function_score = JSONField(
+        verbose_name=_('Function Score'),
+        help_text=_('JSON string containing an elasticsearch function score config.'),
+        null=False,
+        blank=False,
+        default={
+            "functions": [],
```
Comment (clintonb) [inline course_discovery/apps/edx_haystack_extensions/models.py]: Nit: use single quotes.

File: course_discovery/apps/edx_haystack_extensions/models.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+from solo.models import SingletonModel
+
+
+class ElasticsearchBoostConfig(SingletonModel):
+    """
+    Model used to store the elasticsearch boost configuration.
+    This includes a default JSON config for the function_score.
+    """
+
+    # EXAMPLE FUNCTIONS
```
Comment (clintonb) [inline course_discovery/apps/edx_haystack_extensions/models.py]: This, along with an explanation of the effects, would be better housed in `docs/`.

**Comment by clintonb** [comment]
@bderusha what I am seeing looks good. I would prefer to see tests before merging. Docs should also be updated. That's not a hard requirement; but, it would be ideal if that happened sooner (while the information is fresh on your mind).

**Comment by bderusha** [comment]
@clintonb updated based on your comments.  Added tests, docs, and syntax tweaks

**Comment by bderusha** [comment]
The only thing I'm not sure of is how to make the docs test that they render properly

**Comment by clintonb** [comment]
👍

**Pull request merged**